docs/types/reference-types.rst

```diff
@@ -27,20 +27,19 @@ Every reference type has an additional
 annotation, the "data location", about where it is stored. There are three data locations:
 ``memory``, ``storage`` and ``calldata``. Calldata is a non-modifiable,
 non-persistent area where function arguments are stored, and behaves mostly like memory.
-It is required for parameters of external functions but can also be used for other variables.
 
+.. note::
+    If you can, try to use ``calldata`` as a data location because it will avoid copies and
+    also makes sure that the data cannot be modified. Calldata is typically preferred over
+    memory in  external functions (though not required after version 0.6.9). Arrays and
+    structs with ``calldata`` data location can also be returned from functions, but it
+    is not possible to allocate such types.
```

Comment:
I think that the existing note already emphasizes that `calldata` is preferred so the added sentence is a bit redundant. My suggestion would be to leave that note as is (though moving it up is fine) and instead add another one specifically about changes from 0.6.9. What do you think about this:

```suggestion
.. note::
    If you can, try to use ``calldata`` as data location because it will avoid copies and
    also makes sure that the data cannot be modified. Arrays and structs with ``calldata``
    data location can also be returned from functions, but it is not possible to
    allocate such types.

.. note::
    Prior to version 0.6.9 data location for reference-type arguments was limited to
    ``calldata`` in external functions, ``memory`` in public functions and either
    ``memory`` or ``storage`` in internal and private ones.
    Now ``memory`` and ``calldata`` are allowed in all functions regardless of their visibility.
```

Reply:
Sounds good, thanks for all the help on this one!